cmd/flux/cmd/test.go

```diff
@@ -358,29 +401,126 @@ func (s systemfs) Close() error {
 	return nil
 }
 
-func gatherFromDir(filename string) ([]string, fs, error) {
-	var files []string
+type prefixfs struct {
+	prefix string
+	fs     filesystem.Service
+}
+
+func (s prefixfs) Open(fpath string) (filesystem.File, error) {
+	fpath = filepath.Join(s.prefix, fpath)
+	return s.fs.Open(fpath)
+}
+
+func gatherFromDir(filename string) ([]string, fs, edit.TestModules, error) {
+	var (
+		files   []string
+		modules edit.TestModules
+	)
+
+	// Find a test root above the root if it exists.
+	if name, fs, ok, err := findParentTestRoot(filename); err != nil {
+		return nil, nil, nil, err
+	} else if ok {
+		if err := modules.Add(name, fs); err != nil {
+			return nil, nil, nil, err
+		}
+	}
+
 	if err := filepath.Walk(
 		filename,
 		func(path string, info os.FileInfo, err error) error {
 			if isTestFile(info, path) {
 				files = append(files, path)
+			} else if isTestRoot(path) {
+				name, err := readTestRoot(os.Open(path))
+				if err != nil {
+					return err
+				}
+
+				if err := modules.Add(name, prefixfs{
+					prefix: filepath.Dir(path),
+					fs:     systemfs{},
+				}); err != nil {
+					return err
+				}
 			}
 			return nil
 		}); err != nil {
-		return nil, nil, err
+		return nil, nil, nil, err
 	}
-	return files, systemfs{}, nil
+	return files, systemfs{}, modules, nil
 }
 
-func gatherFromFile(filename string) ([]string, fs, error) {
-	return []string{filename}, systemfs{}, nil
+func gatherFromFile(filename string) ([]string, fs, edit.TestModules, error) {
+	var modules edit.TestModules
+
+	// Find a test root above the root if it exists.
+	if name, fs, ok, err := findParentTestRoot(filename); err != nil {
+		return nil, nil, nil, err
+	} else if ok {
+		if err := modules.Add(name, fs); err != nil {
+			return nil, nil, nil, err
+		}
+	}
+	return []string{filename}, systemfs{}, modules, nil
 }
 
 func isTestFile(fi os.FileInfo, filename string) bool {
 	return !fi.IsDir() && strings.HasSuffix(filename, "_test.flux")
 }
 
+const testRootFilename = "fluxtest.root"
+
+func isTestRoot(filename string) bool {
+	return filepath.Base(filename) == testRootFilename
+}
+
+func readTestRoot(f io.Reader, err error) (string, error) {
+	if err != nil {
+		return "", err
+	}
+
+	if rc, ok := f.(io.ReadCloser); ok {
+		defer func() { _ = rc.Close() }()
+	}
+	name, err := ioutil.ReadAll(f)
+	if err != nil {
+		return "", err
+	}
+
+	s := string(bytes.TrimSpace(name))
+	if len(s) == 0 {
+		return "", errors.New(codes.FailedPrecondition, "test module name must be non-empty")
+	}
+	return s, nil
+}
+
+// findParentTestRoot searches the parents to find a test root.
+// This function only works for the system filesystem and isn't meant
+// to be used with archive filesystems.
```

Comment:
This is a good comment. Is there a way we can assert in the function itself that it's being run in that context?

Reply:
I don't think so.